plugins/inputs/mavlink/mavlink.go

```diff
@@ -0,0 +1,196 @@
+			if err != nil {
+				return fmt.Errorf("serial baud rate not valid: %w", err)
+			}
+			baudRate = r
+		}
+
+		s.endpointConfig = []gomavlib.EndpointConf{
+			gomavlib.EndpointSerial{
+				Device: device,
+				Baud:   baudRate,
+			},
+		}
+	case "tcp":
+		if len(host) > 0 {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointTCPClient{
+					Address: host + ":" + port,
+				},
+			}
+		} else {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointTCPServer{
+					Address: ":" + port,
+				},
+			}
+		}
+
+	case "udp":
+		if len(host) > 0 {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointUDPClient{
+					Address: host + ":" + port,
+				},
+			}
+		} else {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointUDPServer{
+					Address: ":" + port,
+				},
+			}
+		}
+
+	default:
+		return fmt.Errorf("could not parse url %s", s.URL)
+	}
+
+	// Compile filter
+	s.filter, err = filter.Compile(s.FilterPattern)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Mavlink) Start(acc telegraf.Accumulator) error {
+	// Start MAVLink endpoint
+	connection, err := gomavlib.NewNode(gomavlib.NodeConf{
+		Endpoints:              s.endpointConfig,
+		Dialect:                ardupilotmega.Dialect,
+		OutVersion:             gomavlib.V2,
+		OutSystemID:            s.SystemID,
+		StreamRequestEnable:    s.StreamRequestEnable,
+		StreamRequestFrequency: s.StreamRequestFrequency,
+	})
+	if err != nil {
+		return &internal.StartupError{
+			Err:   fmt.Errorf("connecting to mavlink endpoint %s failed: %w", s.URL, err),
+			Retry: true,
+		}
+	}
+	s.connection = connection
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	s.cancel = cancelFunc
+
+	// Start routine to connect to Mavlink and stream out data async
+	go func(ctx context.Context) {
+		defer s.connection.Close()
+		if ctx.Err() != nil {
+			return
+		}
+
+		// Process MAVLink messages
+		// Use reflection to retrieve and handle all message types.
+		// (There are several hundred Mavlink message types)
+		for evt := range s.connection.Events() {
+			if ctx.Err() != nil {
+				return
+			}
+			switch evt := evt.(type) {
+			case *gomavlib.EventFrame:
+				result := convertEventFrameToMetric(evt, s.filter)
+
+				if result != nil {
+					result.AddTag("source", s.URL)
+					acc.AddMetric(result)
+				}
+
+			case *gomavlib.EventChannelOpen:
+				s.Log.Debugf("Mavlink channel opened")
+
+			case *gomavlib.EventChannelClose:
+				s.Log.Debugf("Mavlink channel closed")
+			}
+		}
+	}(ctx)
+
+	return nil
+}
+
+func (*Mavlink) Gather(telegraf.Accumulator) error {
+	return nil
+}
+
+func (s *Mavlink) Stop() {
+	s.cancel()
+}
+
+func init() {
+	inputs.Add("mavlink", func() telegraf.Input {
+		return &Mavlink{
+			URL:                    "udp://:14540",
+			FilterPattern:          make([]string, 0),
+			SystemID:               254,
+			StreamRequestEnable:    true,
+			StreamRequestFrequency: 4,
```

Comment:
Setting default values should be done in `Init()`! Here you only need to keep values where the `nil` value is a valid setting.
```suggestion
			SystemID:               254,
			StreamRequestEnable:    true,
```

Reply:
Changed this so the only value that is set here is `SystemID`, for which `0` is a valid value but we want the default to be `254` to not interfere with other components using ID 0.